File: FirstAndLastElementInSortedArray.java
// Time Complexity : O(log(n)) Binary search applied for searching the element in array
// Space Complexity : O(1) no extra space used other than constants
// Did this code successfully run on Leetcode : YES
// Any problem you faced while coding this : NO

public class FirstAndLastElementInSortedArray {
    public int[] searchRange(int[] nums, int target) {
        int [] result = new int[]{-1,-1};

        //base case
        if(nums == null || nums.length == 0) {
            return result;
        }
        int first = binarySearchFirst(nums, target);
        if(first == -1){
            return new int[] {-1,-1};
        }
        int second = binarySearchSecond(nums, target, first);
        return new int[]{first, second};
    }

    private int binarySearchFirst(int[] nums, int target) {
        int low = 0;
        int high = nums.length -1;
        while(low <= high){
            int mid = low + (high - low)/2;
            if(nums[mid] == target) {
                if((mid == low) || (nums[mid-1] != nums[mid])){
                    return mid;
                }
                high =  mid - 1;
            }
            else if (nums[mid] > target) {
                high = mid - 1;
            }
            else{
                low = mid + 1;
            }
        }
        return -1;
    }

    private int binarySearchSecond(int[] nums, int target, int low) {
        int high = nums.length -1;

        while(low <= high) {
            int mid = low + (high - low)/2;
            if(nums[mid] == target) {
                if((mid == high) || (nums[mid] != nums[mid + 1])){
                    return mid;
                }
                low = mid + 1;
            }
            else if (nums[mid] > target) {
                high = mid - 1;
            }
            else{
                low = mid + 1;
            }
        }
        return -1;
    }
}

File: first_and_last_element_in_sorted_array.py
# Time Complexity: O(log(n))
# Space Complexity: O(1)
class Solution:
    def binarysearchleft(self, nums: List[int], target: int) ->int:
        low=0
        high=len(nums)-1
        while low<=high:
            mid= low+(high-low)//2
            if nums[mid]==target:
                if mid==0 or nums[mid-1]<nums[mid]:
                    return mid
                else:
                    high=mid-1
            elif nums[mid]<target:
                low=mid+1
            else:
                high=mid-1
        return -1
    def binarysearchright(self, nums: List[int], target: int) ->int:
        low=lowright
        high=len(nums)-1
        while low<=high:
            mid= low+(high-low)//2
            if nums[mid]==target:
                if mid==len(nums)-1 or nums[mid+1]>nums[mid]:
                    return mid
                else:
                    low=mid+1
            elif nums[mid]<target:
                low=mid+1
            else:
                high=mid-1
        return -1
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        if nums==None or len(nums)==0: return [-1,-1]
        if nums[0]>target or nums[len(nums)-1]<target: return [-1,-1]
        global lowright
        left=self.binarysearchleft(nums,target)
        if left==-1:
            return [-1,-1]
        lowright=left
        right=self.binarysearchright(nums,target)
        return [left,right]
        